File: lowestCommonAncestorofBinaryTree.py
#TimeComplexity:O(N) 
#SpaceComplexity: O(Depth)
#Did this code successfully run on Leetcode : Yes 
#Any problem you faced while coding this : No
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root==None:
            return None
        if root==p or root==q:#if root is wither p or q return root
            return root
        
        left=self.lowestCommonAncestor(root.left,p,q)#we will get a value from left if p or q exists else we will get None
        right=self.lowestCommonAncestor(root.right,p,q)# similarly for right
        
        if left!=None and right!=None:#Finally if both are not null it means root is the answer as each of p, q are present to either sides of root
            return root
        elif left!=None:#Here both are present to the left hence only left is not none and right is None
            return left
        else: #else they are present to the right side
            return right
        

File: Lowest_common_ancestor_of_Binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<TreeNode> path1;
    List<TreeNode> path2;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        path1 = new ArrayList<>();
        path2 = new ArrayList<>();
        backtrack(root, p, path1);
        backtrack(root, q, path2);
        for(int i = 0; i<path1.size(); i++){
            if(path1.get(i) != path2.get(i)){
                return path2.get(i - 1);
            }
        }
        return null;
    }
    
    private void backtrack(TreeNode root, TreeNode p, List<TreeNode> path){
        //base
        if(root == null) return;
        if(root == p){ path.add(root);
        path.add(root);
        return;
        }
                      
        
        //logic
        //action
        path.add(root);
        //recurse
        backtrack(root.left, p, path);
        backtrack(root.right, p, path);
        if(path.get(path.size() - 1) == p) return;
        
        //backtrack
        path.remove(path.size() -1);
    }
}

//TC: O(N)
//SC: O(1)
